src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp

```diff
@@ -44,27 +44,60 @@
-    __ subptr(rsp, 64);
-    __ movdbl(Address(rsp, 0), xmm0);
-    __ movdbl(Address(rsp, 8), xmm1);
-    __ movdbl(Address(rsp, 16), xmm2);
-    __ movdbl(Address(rsp, 24), xmm3);
-    __ movdbl(Address(rsp, 32), xmm4);
-    __ movdbl(Address(rsp, 40), xmm5);
-    __ movdbl(Address(rsp, 48), xmm6);
-    __ movdbl(Address(rsp, 56), xmm7);
+  if (UseSSE >= 2) {
+    const int xmm_size = wordSize * LP64_ONLY(2) NOT_LP64(4);
+    __ subptr(rsp, xmm_size * 8);
+    __ movdbl(Address(rsp, xmm_size * 0), xmm0);
+    __ movdbl(Address(rsp, xmm_size * 1), xmm1);
+    __ movdbl(Address(rsp, xmm_size * 2), xmm2);
+    __ movdbl(Address(rsp, xmm_size * 3), xmm3);
+    __ movdbl(Address(rsp, xmm_size * 4), xmm4);
+    __ movdbl(Address(rsp, xmm_size * 5), xmm5);
+    __ movdbl(Address(rsp, xmm_size * 6), xmm6);
+    __ movdbl(Address(rsp, xmm_size * 7), xmm7);
+  } else if (UseSSE >= 1) {
+    const int xmm_size = wordSize * LP64_ONLY(1) NOT_LP64(2);
+    __ subptr(rsp, xmm_size * 8);
+    __ movflt(Address(rsp, xmm_size * 0), xmm0);
+    __ movflt(Address(rsp, xmm_size * 1), xmm1);
+    __ movflt(Address(rsp, xmm_size * 2), xmm2);
+    __ movflt(Address(rsp, xmm_size * 3), xmm3);
+    __ movflt(Address(rsp, xmm_size * 4), xmm4);
+    __ movflt(Address(rsp, xmm_size * 5), xmm5);
+    __ movflt(Address(rsp, xmm_size * 6), xmm6);
+    __ movflt(Address(rsp, xmm_size * 7), xmm7);
+  } else {
+    // There should be nothing to save.
+  }
 }
 
 static void restore_xmm_registers(MacroAssembler* masm) {
-    __ movdbl(xmm0, Address(rsp, 0));
-    __ movdbl(xmm1, Address(rsp, 8));
-    __ movdbl(xmm2, Address(rsp, 16));
-    __ movdbl(xmm3, Address(rsp, 24));
-    __ movdbl(xmm4, Address(rsp, 32));
-    __ movdbl(xmm5, Address(rsp, 40));
-    __ movdbl(xmm6, Address(rsp, 48));
-    __ movdbl(xmm7, Address(rsp, 56));
-    __ addptr(rsp, 64);
+  if (UseSSE >= 2) {
+    const int xmm_size = wordSize * LP64_ONLY(2) NOT_LP64(4);
+    __ movdbl(xmm0, Address(rsp, xmm_size * 0));
+    __ movdbl(xmm1, Address(rsp, xmm_size * 1));
+    __ movdbl(xmm2, Address(rsp, xmm_size * 2));
+    __ movdbl(xmm3, Address(rsp, xmm_size * 3));
+    __ movdbl(xmm4, Address(rsp, xmm_size * 4));
+    __ movdbl(xmm5, Address(rsp, xmm_size * 5));
+    __ movdbl(xmm6, Address(rsp, xmm_size * 6));
+    __ movdbl(xmm7, Address(rsp, xmm_size * 7));
+    __ addptr(rsp, xmm_size * 8);
+  } else
+  if (UseSSE >= 1) {
+    const int xmm_size = wordSize * LP64_ONLY(1) NOT_LP64(2);
+    __ movflt(xmm0, Address(rsp, xmm_size * 0));
+    __ movflt(xmm1, Address(rsp, xmm_size * 1));
+    __ movflt(xmm2, Address(rsp, xmm_size * 2));
+    __ movflt(xmm3, Address(rsp, xmm_size * 3));
+    __ movflt(xmm4, Address(rsp, xmm_size * 4));
+    __ movflt(xmm5, Address(rsp, xmm_size * 5));
+    __ movflt(xmm6, Address(rsp, xmm_size * 6));
+    __ movflt(xmm7, Address(rsp, xmm_size * 7));
+    __ addptr(rsp, xmm_size * 8);
+  } else {
+    // There should be nothing to save.
+  }
 }
```

Comment:
Take a look at methodHandles_x86.cpp, there is pushing/popping code like this:
#ifdef _LP64
  __ movdbl(Address(rsp, 0), xmm0);
#else
  if  (UseSSE >= 2) {
    __ movdbl(Address(rsp, 0), xmm0);
  } else if (UseSSE == 1) {
    __ movflt(Address(rsp, 0), xmm0);
  } else {
    __ fst_d(Address(rsp, 0));
  }
#endif // LP64

IOW, it also has a branch for no SSE at all. :-)

BTW, I am almost certain that we only ever need to save/restore xmm0. The relevant code is (afaict) only ever called by the interpreter (which only uses xmm0) and methodHandles (which also only seems to care about xmm0 - see the various code sequences there 'save FP result'). Grep for load_heap_oop(), that gives all relevant entries. But that is outside the scope of this patch.

Reply:
To extend a little bit on that last comment: we introduced that xmm save/restore code when we still had RBs and WBs, and we needed a WB in front of C1's CAS-obj. That would mess with C1 keeping FP values in xmm registers across the then-WB. With the move to LRB, this was no longer necessary, and all remaining calls into the asm LRB routines are from the interpreter and thus only need to save/restore xmm0 (which could reasonably be folded into save/restore of other registers and save one add/sub each). The C1 LRB stub calls C1_MacroAssembler::save_live_registers_no_oop_map() which is already smart about which registers to save in context of C1.